src/vs/platform/quickinput/browser/quickInput.ts

```diff
@@ -227,4 +230,75 @@ export class QuickInputService extends Themable implements IQuickInputService {
 			})
 		};
 	}
+
+	formatPinnedItems(storageKey: string, quickPick: IQuickPick<IQuickPickItem>, storageService: IStorageService, callback?: () => Promise<void>): void {
+		quickPick.onDidTriggerItemButton((e) => {
+			if (e.button.iconClass === ThemeIcon.asClassName(Codicon.pin) || e.button.iconClass !== ThemeIcon.asClassName(Codicon.pinned)) {
+				this._formatPinnedItems(storageKey, quickPick, storageService, callback, e);
+			}
+		});
+		this._formatPinnedItems(storageKey, quickPick, storageService, callback);
+	}
+
+	private async _formatPinnedItems(storageKey: string, quickPick: IQuickPick<IQuickPickItem>, storageService: IStorageService, callback?: () => Promise<void>, event?: IQuickPickItemButtonEvent<IQuickPickItem>): Promise<void> {
+		const formattedItems: (IQuickPickItem | IQuickPickSeparator)[] = [];
+		const labels = this._getPinnedLabels(storageKey, storageService);
+		const pinnedLabels = event ? this._updatePinned(storageKey, event.item.label, storageService, new Set(labels).has(event.item.label)) : labels;
+		for (const label of pinnedLabels) {
+			const item = quickPick.items.find(i => i.label === label);
+			if (item) {
+				const pinnedItem = Object.assign({}, item);
+				this._updateButtons(pinnedItem, false);
+				formattedItems.push(pinnedItem);
+			}
+		}
+		if (pinnedLabels.length) {
+			formattedItems.push({ type: 'separator', label: localize("terminal.commands.pinned", 'Pinned') });
+		}
+
+		for (const item of quickPick.items.filter(i => !!i.label)) {
+			this._updateButtons(item, true);
+			formattedItems.push(item);
+		}
```

Comment:
The `Set` could be used here to track which items are filtered already so only a single item matching a label is moved to pinned. Do we need all original items as pinned items will get ignored in search?

Reply:
yes we need all items. the filter there is to ensure it's not a quick-pick separator. 

i'll use a set above though to determine which should be pinned